keps/sig-network/0007-pod-ready++.md

```diff
@@ -0,0 +1,173 @@
+
+For example, during deployment rolling update, a new pod becomes ready. On the other hand, service, network policy and load-balancer are not yet ready for the new pod due to whatever reason (e.g. slowness in api machinery, endpoints controller, kube-proxy, iptables or infrastructure programming). This may cause service disruption or lost of backend capacity. In extreme cases, if rolling update completes before any new replacement pod actually start serving traffic, this will cause service outage. 
+
+
+### Goals
+
+- Allow extra signals for pod readiness.
+
+### Non-Goals
+
+- Provide generic framework to solve all transition problems in k8s (e.g. blue green deployment).
+
+## Proposal
+
+[K8s Proposal: Pod Ready++](https://docs.google.com/document/d/1VFZbc_IqPf_Msd-jul7LKTmGjvQ5qRldYOFV0lGqxf8/edit#)
+
+### PodSpec
+Introduce an extra field called ReadinessGates in PodSpec. The field stores a list of ReadinessGate structure as follows: 
+```yaml
+type ReadinessGate struct {
+	conditionType string	
+}
+```
+The ReadinessGate struct has only one string field called ConditionType. ConditionType refers to a condition in the PodCondition list in PodStatus. And the status of conditions specified in the ReadinessGates will be evaluated for pod readiness. If the condition does not exist in the PodCondition list, its status will be default to false. 
+
+#### Constraints:
+- ReadinessGates can only be specified at pod creation. 
+- No Update allowed on ReadinessGates.
+- ConditionType must conform to the naming convention of custom pod condition.
+
+### Pod Readiness
+Change the pod readiness definition to as follows:
+```
+Pod is ready == containers are ready AND conditions in ReadinessGates are True
+```
+Kubelet will evaluate conditions specified in ReadinessGates and update the pod “Ready” status. For example, in the following pod spec, two readinessGates are specified. The status of “www.example.com/feature-1” is false, hence the pod is not ready. 
+
+```yaml
+Kind: Pod 
+… 
+spec: 
+  readinessGates:
+  - conditionType: www.example.com/feature-1
+  - conditionType: www.example.com/feature-2
+… 
+status:
+  conditions:
+  - lastProbeTime: null
+    lastTransitionTime: 2018-01-01T00:00:00Z
+    status: "False"
+    type: Ready
+  - lastProbeTime: null
+    lastTransitionTime: 2018-01-01T00:00:00Z
+    status: "False"
+    type: www.example.com/feature-1
+  - lastProbeTime: null
+    lastTransitionTime: 2018-01-01T00:00:00Z
+    status: "True"
+    type: www.example.com/feature-2
+  containerStatuses:
+  - containerID: docker://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+    ready : true
+… 
+```
+
+Another pod condition `ContainerReady` will be introduced to capture the old pod `Ready` condition. 
+```
+ContainerReady is true == containers are ready
```

Comment:
Will this replace the old Ready condition or in addition. ? It has to be in addition otherwise this is a breaking change